### ISSUE TITLE
AMDGPUpro 17.30 Codegen: redundant v_trunc after v_rndne ?

### ISSUE THREAD

**Issue by preda** [RadeonOpenCompute/ROCm#210]
On Ubuntu 17.04, AMDGPU-pro 17.30, Vega64.

I see this sequence generated:
```
v_rndne_f64_e32 v[10:11], v[10:11]                   // 00000001041C: 7E14330A
v_trunc_f64_e32 v[10:11], v[10:11]                   // 000000010420: 7E142F0A
```
Corresponding to *rint(double)* in the context of:
```
long toLong(double x) { return rint(x); }
```

My understanding is that *rndne* (round-to-nearest-even) already produces an integral value. So why is the *trunc* needed afterwards?

**Comment by gstoner** [comment]
@preda We are looking at this.

**Comment by gstoner** [comment]
Please look at 17.40 beta build  http://support.amd.com/en-us/kb-articles/Pages/AMDGPU-Pro-Beta-Mining-Driver-for-Linux-Release-Notes.aspx

**Comment by preda** [comment]
This is still present with ROCm 1.6-180, see e.g.
```
	v_rndne_f64_e32 v[10:11], v[10:11]                         // 0000000087F0: 7E14330A
	v_rndne_f64_e32 v[12:13], v[12:13]                         // 0000000087F4: 7E18330C
	v_trunc_f64_e32 v[10:11], v[10:11]                         // 0000000087F8: 7E142F0A
	v_trunc_f64_e32 v[12:13], v[12:13]                         // 0000000087FC: 7E182F0C
```

**Comment by preda** [comment]
And let me also hint to a possible cause:
there is a pre-packaged code block that does the conversion double -> long. That block starts with a v_trunc_f64.

The rint() already just did a v_rndne_f64, but the compiler fails to optimize away the v_trunc at the beginning of the pre-packaged conversion block.

**Comment by preda** [comment]
I attach ISA for the example kernel above.
[foo.txt](https://github.com/RadeonOpenCompute/ROCm/files/1426806/foo.txt)

**Comment by Srinivasuluch** [comment]
Thank you Mihai for the details.

**Comment by Srinivasuluch** [comment]
Hi Mihai - The issue has been fixed with latest compiler changes, the fix would be available in next ROCm release.
Thank you for posting us the details for the bug.

**Comment by preda** [comment]
Thanks. How can I try the fix out?

**Comment by gstoner** [comment]
It will be  it ROCm 1.7,    One thing I talking to the the team about is have a binary train of  GA( Stable) , Beta( Dev-Stable), Alpha( Canary) via our repo server.

**Comment by acmeman925** [comment]
Hi @preda could this issue be closed?

**Comment by preda** [comment]
Yes, seems fixed in 1.7.
